AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCoreDataClient.m

```diff
@@ -0,0 +1,235 @@
+        NSDictionary *dictionary = [document serializeToDictionary];
+        NSString *jsonDocument;
+        NSError *jsonError;
+        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dictionary options:0 error:&jsonError];
+        if (!error) {
+          jsonDocument = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+        } else {
+          NSString *message = @"Error serializing document for local storage";
+          MSLogError([MSDataStore logTag], message);
+          completionHandler([[MSDocumentWrapper alloc] initWithDataStoreErrorCode:MSACDataStoreLocalStoreError
+                                                                     errorMessage:message
+                                                                       documentId:documentId]);
+          return;
+        }
+
+        // Create a deleted document record.
+        MSDocumentWrapper *createdOrUpdatedDocument = [[MSDocumentWrapper alloc] initWithDeserializedValue:document
+                                                                                                 jsonValue:jsonDocument
+                                                                                                 partition:token.partition
+                                                                                                documentId:documentId
+                                                                                                      eTag:cachedDocument.eTag
+                                                                                           lastUpdatedDate:cachedDocument.lastUpdatedDate
+                                                                                          pendingOperation:operation
+                                                                                                     error:nil];
+
+        // Update local store and return document.
+        [self updateLocalStore:token
+            currentCachedDocument:cachedDocument
+                newCachedDocument:createdOrUpdatedDocument
+                 deviceTimeToLive:deviceTimeToLive
+                        operation:operation];
+        completionHandler(createdOrUpdatedDocument);
+      }
+    }
+  });
+}
+
+/**
+ * Validate an operation.
+ *
+ *@param operation The operation.
+ *
+ *@return YES if the operation is valid; NO otherwise.
+ */
++ (BOOL)isValidOperation:(NSString *)operation {
+  return operation == nil || [kMSPendingOperationCreate isEqualToString:operation] ||
+         [kMSPendingOperationReplace isEqualToString:operation] || [kMSPendingOperationDelete isEqualToString:operation];
+}
+
+/**
+ * Returns a flag indicating if a remote operation should be attempted.
+ *
+ * @param cachedDocument The current cached document (if any).
+ *
+ * @return YES if a remote operation should be attempted; NO otherwise.
+ */
+- (BOOL)needsRemoteOperation:(MSDocumentWrapper *)cachedDocument {
```

Comment:
Is this used in more than one place? If not, let's not obfuscate this logic by putting it in a helper method.

Reply:
It's kind of going against what we are thinking of doing to address @Jamminroot's comments. 

I usually try not to introduce methods called once only, but at the same time, here I think it makes everything (the break down and documentation) more palatable.